Basic operators, math (#247)
javascript-tutorial/pt.javascript.info: 1-js/02-first-steps/08-operators/article.md

> # Operadores básicos, matemática

Reviewer: Faltou traduzir os arquivos `taks.md` e `solution.md` pertinente aos exemplos existentes.

Poderia adicioná-los na tradução, por gentileza?

- 1-increment-order
- 2-assignment-result
- 3-primitive-conversions-questions
- 4-fix-prompt

<img width="1111" alt="image" src="https://user-images.githubusercontent.com/35238611/184464974-8e91a902-7615-4d22-a7ca-610f50a28327.png">

Author: > Faltou traduzir os arquivos taks.md e solution.md pertinente aos exemplos existentes.
> 
> Poderia adicioná-los na tradução, por gentileza?

Claro que sim! 😀 
Obrigado por me lembrar destes arquivos! Vou começar agora a trabalhar neles!

Reviewer: Revisei as alterações, tudo em ordem, com um pequeno detalhe que não faz parte da tradução ...

O nome da chave `inportance` no começo dos arquivos `task.md` deve ser mantido, pois o conteúdo acima  de `---` provavelmente é utilizado pela plataforma para alguma operação ou estruturação necessária.

> Não traduzir as chaves de identificação dos **metadados**, pois não fazem parte do conteúdo do artigo, mas sim para fins do projeto ["server"](https://github.com/javascript-tutorial/server) que construirá o site "https://javascript.info"

Por gentileza, retroceder a alteração da chave `importance` em todos os arquivos:

1-increment-order
2-assignment-result
3-primitive-conversions-questions
4-fix-prompt

- Um operador é "unário" se este tiver um único operando. Por exemplo, a negação unária `-` inverte o sinal de um número:

    ```js run
    let x = 1;

    *!*
    x = -x;
    */!*
    alert( x ); // -1, negação unária foi aplicada
    ```
- Um operador é *binário* se este possuir dois operandos. O mesmo menos também existe em forma binária:

    ```js run no-beautify
    let x = 1, y = 3;
    alert( y - x ); // 2, menos binário subtrai valores
    ```

    Formalmente, nos exemplos acima nós temos dois operadores diferentes que compartilham o mesmo símbolo: o operador de negação, um operador unário que inverte o sinal, e o operador de subtração, um operador binário que subtrai um número de outro.

## Matemática

As seguintes operações matemáticas são suportadas:

- Adição `+`,
- Subtração `-`,
- Multiplicação `*`,
- Divisão `/`,
- Resto `%`,
- Exponenciação `**`.

Os quatro primeiros são diretos, enquanto que `%` e `**` precisam de alguma explicação a respeito.

### Resto %

O operador de resto `%`, apesar de sua aparência, não possui relação com porcentagens.

> O resultado de `a % b`é o [resto](https://pt.wikipedia.org/wiki/Resto_da_divis%C3%A3o_inteira) da divisão inteira de `a` por `b`.

Author: Muito obrigado pela correção! Eu a apliquei aqui: [Typo-fix: space inserted between letters](https://github.com/javascript-tutorial/pt.javascript.info/pull/247/commits/4e7ed7f6c1ae5a3572fd3b202a2dd6c426b0041d).

```js run
alert('1' + 2 + 2); // "122" e não "14"
```
Aqui o primeiro operando é uma string, com o compilador tratando os outros dois operandos também como strings. O `2` é concatenado ao `'1'`, ficando `'1' + 2 = "12"` e `"12" + 2 = "122"`.

O binário `+` é o único operador que suporta strings desta forma. Outros operadores aritméticos funcionam apenas com números e sempre convertem seus operandos em números.

Aqui está a demonstração para subtração e divisão:

```js run
alert( 6 - '2' ); // 4, converte '2' em um número
alert( '6' / '2' ); // 3, converte ambos os operandos em números
```

## Conversão numérica, + unário

O mais `+` existe em duas formas: na forma binária que usamos acima e na forma unária.

O positivo unário ou, em outras palavras, o operador mais `+` aplicado a um valor único, não faz nada com números. Mas se o operando não é um número, o positivo unário o converte em número.

Por exemplo:

```js run
// Sem efeito em números
let x = 1;
alert( +x ); // 1

let y = -2;
alert( +y ); // -2

*!*
// Converte não-numéricos
alert( +true ); // 1
alert( +"" );   // 0
*/!*
```

> Isto na verdade faz a mesma coisa que `Number(...)`, mas de forma mais curta.

Author: Muito obrigado pelas observações. Eu apliquei as sugestões/correções aqui: [Grammatical improvements](https://github.com/javascript-tutorial/pt.javascript.info/pull/247/commits/b3c65d6f32bbd2f0f5489aa63dba5b4d067e7d43).

A necessidade de converter strings em números surge com frequência. Por exemplo, se nós estivermos pegando valores de campos de formulário HTML, eles são normalmente strings. E se nós quisermos somá-los?

O positivo binário os adicionaria como strings:

```js run
let apples = "2";
let oranges = "3";

alert( apples + oranges ); // "23", o binário positivo concatena strings
```

Se nós quisermos tratá-los como números, nós precisamos convertê-los e então somá-los:

```js run
let apples = "2";
let oranges = "3";

*!*
// ambos os valores convertidos para números antes do binário positivo
alert( +apples + +oranges ); // 5
*/!*

// a variante mais longa
// alert( Number(apples) + Number(oranges) ); // 5
```

Do ponto de vista matemático, a abundância de sinais de mais pode parecer estranha. Mas de um ponto de vista de um programador, não há nada de especial: positivos unários são aplicados primeiro, eles convertem strings em números, e então o positivo binário os soma.

> Por que positivos unários são aplicados a valores antes dos positivos binários? Como iremos ver, é por causa da *maior precedência* destes.

Author: Muito obrigado pelas observações. Eu apliquei as sugestões/correções aqui: [Grammatical improvements](https://github.com/javascript-tutorial/pt.javascript.info/pull/247/commits/b3c65d6f32bbd2f0f5489aa63dba5b4d067e7d43).

alert( c ); // 0
```

> No exemplo acima, o resultado da expressão `(a = b + 1)` é o valor que é atribuído à `a` (isto é, `3`). Este é então usado para avaliações adicionais.

Author: Muito obrigado pelas observações. Eu apliquei as sugestões/correções aqui: [Grammatical improvements](https://github.com/javascript-tutorial/pt.javascript.info/pull/247/commits/b3c65d6f32bbd2f0f5489aa63dba5b4d067e7d43).

alert(a); // *!*2*/!*
```

Na linha `(*)`, a forma *prefixa* `++counter` incrementa `counter` e retorna o novo valor, `2`. Então, o `alert` mostra `2`.

Agora, vamos usar a forma pós-fixa:

```js run
let counter = 1;
let a = counter++; // (*) mudou ++counter para counter++

alert(a); // *!*1*/!*
```

> Na linha `(*)`, a forma pós-fixa `counter++` também incrementa `counter` mas retorna o valor *antigo* (anterior ao incremento). Então, o `alerta` mostra `1`.

Reviewer: **Sugestão de alteração**:

- Adicionar a virgula antes da locução `mas`, alterando `counter mas` para `counter, mas`

<img width="624" alt="image" src="https://user-images.githubusercontent.com/35238611/184464297-5b11b6e6-cb51-4da2-8e01-d1b84500e7eb.png">

> Regra "Locuções entre vírgulas: portanto, por exemplo, na verdade"
>
> https://community.languagetool.org/rule/show/VERB_COMMA_CONJUNCTION?lang=pt

Author: Muito obrigado pelas observações. Eu apliquei as sugestões/correções aqui: [Grammatical improvements](https://github.com/javascript-tutorial/pt.javascript.info/pull/247/commits/b3c65d6f32bbd2f0f5489aa63dba5b4d067e7d43).

Resumindo:

- Se o resultado do incremento/decremento não é usado, não há diferença em qual forma usar:

    ```js run
    let counter = 0;
    counter++;
    ++counter;
    alert( counter ); // 2, as linhas acima fizeram a mesma coisa
    ```
- Se quisermos aumentar o valor *e* imediatamente usar o resultado do operador, nós precisamos da forma prefixa:

    ```js run
    let counter = 0;
    alert( ++counter ); // 1
    ```

> - Se quisermos incrementar um valor mas usar o seu valor anterior, nós precisamos da forma pós-fixa:

Reviewer: **Sugestão de alteração**:

- Adicionar a virgula antes da locução `mas`, alterando `valor mas` para `valor, mas`

<img width="784" alt="image" src="https://user-images.githubusercontent.com/35238611/184464554-99f86f24-2b85-489d-85e7-2bfb69f73101.png">

> Regra "Locuções entre vírgulas: portanto, por exemplo, na verdade"
>
> https://community.languagetool.org/rule/show/VERB_COMMA_CONJUNCTION?lang=pt

Author: Muito obrigado pelas observações. Eu apliquei as sugestões/correções aqui: [Grammatical improvements](https://github.com/javascript-tutorial/pt.javascript.info/pull/247/commits/b3c65d6f32bbd2f0f5489aa63dba5b4d067e7d43).

- RIGHT SHIFT ( `>>` )
- ZERO-FILL RIGHT SHIFT ( `>>>` )

> Estes operadores são muito raramente usados, quando nós precisamos mexer com números no nível mais baixo (bit a bit). Nós não precisaremos destes operadores em nenhum momento, já que desenvolvimento web faz pouco uso deles, mas em algumas áreas específicas, como em criptografia, eles são úteis. Você pode ler o capítulo [Operadores bit a bit](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Guide/Expressions_and_Operators#operadores_bit_a_bit) na MDN quando surgir a necessidade.

Reviewer 2: ```suggestion
Estes operadores são muito raramente usados, e acontece quando nós precisamos mexer com números no nível mais baixo (bit a bit). Nós não precisaremos destes operadores dentro em breve, já que desenvolvimento web faz pouco uso deles, mas em algumas áreas específicas, como em criptografia, eles são úteis. Você pode ler o capítulo [Operadores bit a bit](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Guide/Expressions_and_Operators#operadores_bit_a_bit) na MDN quando surgir a necessidade.
```

## Vírgula

O operador vírgula `,`, é um dos operadores mais raros e incomuns. Às vezes, ele é usado para escrever código mais curto, de modo que precisamos conhecê-lo para entender o que está ocorrendo.

> O operador vírgula nos permite avaliar várias expressões, dividindo-as com uma vírgula `,`. Cada uma delas é avaliada mas somente o resultado da última é retornado.

Reviewer: **Sugestão de alteração**:

- Adicionar a virgula antes da locução `mas`, alterando `avaliada mas` para `avaliada, mas`.

<img width="581" alt="image" src="https://user-images.githubusercontent.com/35238611/184464658-306dffe4-c02b-43fc-b949-77c15fb03532.png">

> Regra "Locuções entre vírgulas: portanto, por exemplo, na verdade"
>
> https://community.languagetool.org/rule/show/VERB_COMMA_CONJUNCTION?lang=pt

Author: Muito obrigado pelas observações. Eu apliquei as sugestões/correções aqui: [Grammatical improvements](https://github.com/javascript-tutorial/pt.javascript.info/pull/247/commits/b3c65d6f32bbd2f0f5489aa63dba5b4d067e7d43).

Por exemplo:

```js run
*!*
let a = (1 + 2, 3 + 4);
*/!*

alert( a ); // 7 (o resultado de 3 + 4)
```

Aqui, a primeira expressão `1 + 2` é avaliada e o seu resultado é descartado. Então, `3 + 4` é avaliado e retornado como resultado.

```smart header="Vírgula tem uma precedência muito baixa"
Por favor observe que o operador vírgula tem precedência muito baixa, mais baixa que `=`, de modo que parênteses são importantes no exemplo acima.

Sem eles: `a = 1 + 2, 3 + 4` avalia `+` primeiro, somando os números e resultando em `a = 3, 7`, com o operador de atribuição `=` atribuindo `a = 3`, e ignorando o restante. É como `(a = 1 + 2), 3 + 4`.
```

Por que precisamos de um operador que descarta tudo exceto a última expressão?

> Às vezes, as pessoas o usam e construtos mais complexos para pôr várias ações em uma linha.

Reviewer 2: ```suggestion
Às vezes, as pessoas o usam em construções mais complexas para pôr várias ações em uma linha.
```

Por exemplo:

```js
// três operações em uma linha
for (*!*a = 1, b = 3, c = a * b*/!*; a < 10; a++) {
 ...
}
```

> Tais truques são usados em muitos frameworks JavaScript. É por isto que nós os estamos mencionando. Mas geralmente eles não melhoram a legibilidade do código de modo que nós precisamos pensar bem antes de usá-los.

Reviewer 2: ```suggestion
Tais truques são usados em muitos frameworks JavaScript. É por isto que nós os estamos mencionando. Mas geralmente eles não melhoram a legibilidade do código, de modo que nós precisamos pensar bem antes de usá-los.
```
